### ISSUE TITLE
Highlight selectedText cuts off the last selected line, and jumpToHighlightArea jumps to the bottom of the selectionRegion instead of the top

### ISSUE THREAD

**Issue by lucasgadams** [react-pdf-viewer/react-pdf-viewer#1294]
I'm encountering 2 issues using the HighlightPlugin. What I am essentially trying to do is get the selected text from a user, do some processing on that, and then create a highlight bookmark of sorts that displays the text and will jump to the destination when clicked. These issues appear on all the pdfs i've tried, including the Parameters for Opening PDF Files default.

Issues:
1. The `selectedText` in `RenderHighlightTargetProps` from `renderHighlightTarget` seems to cut off the last selected line. If there are 4 highlighted lines from a user, the selectedText will contain the first 3.
2. I'm storing the `selectedText` and the `selectionRegion`, and then displaying them to a user in the side panel. When clicked, I use the `jumpToHighlightArea` to scroll to the spot of the highlight. This is always off for me, and it looks to me like it is actually scrolling just to the bottom of the `selectedText` area, so that is always just out of view above where the page is scrolled to.

**Comment by lucasgadams** [comment]
Also something else i noticed is every `HighlightArea` i get seems to have the same `height` even when the rectangles should be different sizes. Here is two logged areas that should be different sizes (and if I understand correctly larger than ~1.6):

<img width="273" alt="image" src="https://user-images.githubusercontent.com/36425095/198365518-d516c58c-2f3b-4bcc-9460-b0b3f6c1d32c.png">

**Comment by lucasgadams** [comment]
Ok, I believe I have located at least one issue here https://github.com/react-pdf-viewer/react-pdf-viewer/blob/master/packages/highlight/src/Tracker.tsx#L231 . 

Am I not understanding correctly what the `selectionRegion` is supposed to do? That is what I am trying to use to get the coordinates to scroll the viewer to the highlighted item. Why do we take the last of that array to be the `selectionRegion`?

**Comment by lucasgadams** [comment]
Changing my code to use `highlightProps.highlightAreas[0]` instead of `highlightProps.selectionRegion` fixes the issue of where it scrolls to using `jumpToHighlightArea` so I can use that. I still have the issue that the last line of selected text is cut off and not included in the `highlightProps.selectedText`

**Comment by phuocng** [comment]
> 1. The `selectedText` in `RenderHighlightTargetProps` from `renderHighlightTarget` seems to cut off the last selected line. If there are 4 highlighted lines from a user, the selectedText will contain the first 3.

Regarding this issue, I can't reproduce the issue. Can you fork this [sample project](https://codesandbox.io/s/react-pdf-viewer-highlight-mr92zp) and let me know the steps to reproduce the issue?

![Screenshot 2022-11-05 at 18 19 28](https://user-images.githubusercontent.com/57786711/200117371-b4334c88-aaf7-4162-97a9-da1cc7d2e1a1.png)

**Comment by ryanmearns** [comment]
I'm having the issue. I'll try and reproduce it with the example project.

**Comment by lucasgadams** [comment]
Ok, i think I've located where the issue is happening. I could not repro on the sandbox, and so I stripped everything out of my app until it was bare bones, and eventually found that toggling `reactStrictMode` in `next.config.mjs` off fixed the issue. I am not familiar enough with react/nextjs to understand why this is happening. I have created a minimal repo that shows the issue here https://github.com/lucasgadams/react-pdf-highlight-bug

**Comment by lucasgadams** [comment]
I can also confirm this only happens when running in development mode, so using `next build && next start` will fix the issue.

**Comment by ryanmearns** [comment]
Here is the issue reproduced using NextJs with the issue: https://codesandbox.io/p/github/ryanmearns/test-react-pdf/draft/strange-aryabhata?file=%2FREADME.md

That's good to know that it only happens in Dev mode.

**Comment by lucasgadams** [comment]
On your sandbox (https://codesandbox.io/s/react-pdf-viewer-strict-mode-highlight-hthjjc) I could also not repo:
<img width="1509" alt="image" src="https://user-images.githubusercontent.com/36425095/200615076-fd9f12c2-54e0-44bd-8326-ca36c75ab785.png">

However I could repro with using my [repo in a sandbox](https://codesandbox.io/p/github/lucasgadams/react-pdf-highlight-bug/draft/lively-cherry?file=%2Fnext.config.mjs&workspace=%257B%2522activeFileId%2522%253A%2522cla8dxstz0006l2i27zem9r1r%2522%252C%2522openFiles%2522%253A%255B%2522%252FREADME.md%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522sidekickItems%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522key%2522%253A%2522cla8dy9yt00c63b6hac96llc5%2522%252C%2522isMinimized%2522%253Afalse%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522key%2522%253A%2522cla8dy7xu00853b6hbo2aq0u0%2522%252C%2522isMinimized%2522%253Atrue%257D%255D%257D). When using the same text as you did, I noticed other odd things. It looks like it's not just cutting off the last line, there are other variations of the issue. In the first screen shot below I'm highlighting 2 paragraphs but there is no text actually getting selected. In the second I'm getting text, but you can see that the first line is actually extended past where it is supposed to be based on the cursor, and the last line is also cut off.

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/36425095/200617898-3df23157-7fb5-4bac-b110-49c34b88e678.png">

<img width="1506" alt="image" src="https://user-images.githubusercontent.com/36425095/200618139-8e22bfdc-a483-46a3-aaaa-3de1c6649fe1.png">

**Comment by lucasgadams** [comment]
I also sometimes get errors like this. Not sure how to repro this, it just seems to happen sometimes when clicking a line.
<img width="1507" alt="image" src="https://user-images.githubusercontent.com/36425095/200626370-1acf85b9-f74f-4d08-9434-d666520ea947.png">

**Comment by lucasgadams** [comment]
Actually, the issues with highlighting those 2 paragraphs seem to not be totally fixed by turning off react strict mode (though that does seem to change the behavior still). Here I am selecting the text locally but with strict mode off. You can see only part of the first paragraph is selected:
<img width="1511" alt="image" src="https://user-images.githubusercontent.com/36425095/200627197-23062bca-1c0a-44a6-b740-47571370f6bc.png">

**Comment by lucasgadams** [comment]
I was able to repro one of the new issues identified using your sandbox @phuocng . This one seems to be an issue with highlighting multiple pages:

<img width="1506" alt="image" src="https://user-images.githubusercontent.com/36425095/200632394-a0ae3658-0cdf-4022-9ecd-9c31ffb79948.png">

```
Uncaught RangeError: Invalid array length
    at getRectBetween (highlight.js:369:20)
    at HTMLDivElement.onMouseUpHandler (highlight.js:413:32)
getRectBetween @ highlight.js:369
onMouseUpHandler @ highlight.js:413
```

To be clear, i think this is a different but potentially related issue.

**Comment by lucasgadams** [comment]
Ok I think I've found the cause, I believe it is actually using React 18 vs 17 that will repro the bug, not using nextjs. When running your sandbox @phuocng I noticed this warning:
```
Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot
```

So the demo code is actually running **React 17** not **18**. Here is a [forked sandbox](https://codesandbox.io/s/react-pdf-viewer-strict-mode-highlight-forked-13j5y4?file=/src/index.tsx) from the demo that changes the `index.tsx` to:
```
import React from "react";
import { createRoot } from "react-dom/client";

import App from "./App";

const rootElement = document.getElementById("root");
const root = createRoot(rootElement); // createRoot(container!) if you use TypeScript
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
```


This clear the warning, and then running the code I can repro the original issue with the lines cut off: 
<img width="1509" alt="image" src="https://user-images.githubusercontent.com/36425095/200634872-5dde2081-bf01-4551-be78-7231b2310b88.png">

So to recap:
- there seems to be multiple issues when selecting text, including not getting any text, not selecting across multiple paragraphs, having the entire first line selected when it should be offsetted, dropping the last line.
- These issues only appear when using **React 18**. Some of these issues using React 18 are fixed when turning off Strict mode (dropping the last line particularly), though others are not (selecting across multiple paragraphs).
- Additionally, selecting across multiple pages seems to be broken for React 17 and 18
